### PR TITLE
Gitlab CI: install dependencies with Spack instead of TPLS repo [feature/testing-with-spack]

### PR DESCRIPTION
**Pull request by adrienbernede**
### Summary

This PR aims at managing dependencies installation automatically with Spack, _in CI or in development context_.

The implications are:
- The MFEM Spack package is used to manage the dependencies options, which ensure that the package is maintained.
- Installing a new set of dependencies is made straightforward, really. It also produces a `user.mk` (named differently) that points to those installations.

### Application

This is particularly useful in CI context, and hopefully in development context too.

As a demonstration, I changed the CI to test MFEM in exactly the same configurations as before, but managing dependencies with Spack. The CI implementation is much shorter, and can be modified and completed very easily.

I also added some other configurations to showcase what is possible:
- MFEM with Pumi, Sundials and Petsc to cover more of the examples.
- ~~3 Serac-like configurations~~ Not ready for production yet.

### Implementation details

Spack is driven by Uberenv, a python script. The methodology has been tested and vetted in several other projects, including Serac, RAJA, Umpire, Sundials, in order to streamline CI implementation.

The configuration for quartz, lassen, corona (?), is stored in `scripts/uberenv/spack_packages`. It means that it is versioned with the code, and accessible locally. (Compared to on a website for CEED).

`scripts/uberenv/packages` can receive custom Spack packages. Currently it contains a custom version of MFEM packages. Those changes were necessary to use in Uberenv context. They consist in generating the `user.mk` (named differently) that can be used to configure MFEM with Spack-installed dependencies.

### How it works

- Spack is downloaded, configured for the machine, and tuned with local versions of the packages.
- `spack dev-build -u config <mfem-spec>` is used to install MFEM dependencies, and generate the makefile options `user.mk` (named differently). Spack will not install MFEM, it will stop at `config` phase. Good to know: removing `-u config` would have Spack installing MFEM using the local version of the source!
- The CI is managed through a single, generic script: `scripts/gitlab/build_and_test.sh`. Dependencies are installed, and then MFEM is built and tested in a standard way, *not relying on Spack*. Only a Spack spec is required.

### TODO

- [X] Fix issues with some Spack spec for MFEM. Those should probably work.
- [X] Point to a specific commit in Spack (not a `develop`) for stability purpose. Currently pointing to WIP Spack PR to update MFEM package.
And more...
-  [x] Add documentation

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1832](https://github.com/mfem/mfem/pull/1832) | @adrienbernede | @tzanio | @tzanio + @v-dobrev + @acfisher + @cjvogl + @jamiebramwell | 12/16/20 | 02/04/21 | 02/09/21 | |

### CODE REVIEW

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

File: scripts/gitlab/build_and_test.sh
Comment (tzanio) [inline scripts/gitlab/build_and_test.sh:1]: Instead of introducing a new top-level directory, can this be put under `tests`?

i.e. `scripts/gitlab/*` -> `tests/gitlab/*`

Comment (adrienbernede) [reply scripts/gitlab/build_and_test.sh:1]: Sure! that's an artifact of pervious introduction of uberenv...

File: scripts/gitlab/upstream-permission.patch
Comment (tzanio) [inline scripts/gitlab/upstream-permission.patch:6]: Can't this be in `mfem/mfem-uberenv`?

Comment (adrienbernede) [reply scripts/gitlab/upstream-permission.patch:6]: Yes it can!

**Comment by adrienbernede** [comment]
LGTM

File: tests/gitlab/get_mfem_uberenv.sh
Comment (tzanio) [inline tests/gitlab/get_mfem_uberenv.sh:1]: I know these are just a simple scripts, but please add the copyright header as in the scripts insider `../scripts`. Also, consider dropping the `.sh` for uniformity.

File: tests/gitlab/get_mfem_uberenv.sh
Comment (tzanio) [inline tests/gitlab/get_mfem_uberenv.sh:9]: Remove the extra line?

File: tests/gitlab/get_mfem_uberenv.sh
Comment (tzanio) [inline tests/gitlab/get_mfem_uberenv.sh:10]: Is this expected to be run from the top-level MFEM repo? If so, mention that in the beginning of the file (or add a help option as in `../scripts`).

File: tests/gitlab/generate_spack_upstream.sh
Comment (tzanio) [inline tests/gitlab/generate_spack_upstream.sh:7]: Is this file `tests/gitlab/upstream-permission.patch` missing?

File: tests/gitlab/generate_spack_upstream.sh
Comment (tzanio) [inline tests/gitlab/generate_spack_upstream.sh:10]: make sure this works on both Mac and Linux

**Review by tzanio — CHANGES_REQUESTED** [review]

**Comment by tzanio** [comment]
Additional TODOs:

- [ ] Consider adding a README file in `tests/gitlab` and/or mention those in `CONTRIBUTING.md`
- [ ] Mention in `CHANGELOG`?

File: tests/gitlab/get_mfem_uberenv.sh
Comment (adrienbernede) [reply tests/gitlab/get_mfem_uberenv.sh]: Done

File: tests/gitlab/get_mfem_uberenv.sh
Comment (adrienbernede) [reply tests/gitlab/get_mfem_uberenv.sh]: Done

File: tests/gitlab/get_mfem_uberenv.sh
Comment (adrienbernede) [reply tests/gitlab/get_mfem_uberenv.sh]: Done (not using help function... yet?)

File: tests/gitlab/generate_spack_upstream.sh
Comment (adrienbernede) [reply tests/gitlab/generate_spack_upstream.sh]: Not missing, but the path needed to be change after the patch relocation in mfem-uberenv.

Comment (adrienbernede) [reply tests/gitlab/generate_spack_upstream.sh]: Done

File: tests/gitlab/generate_spack_upstream.sh
Comment (adrienbernede) [reply tests/gitlab/generate_spack_upstream.sh]: Done.

**Review by tzanio — APPROVED** [review]
Thanks @adrienbernede !

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
LGTM

**Comment by adrienbernede** [comment]
@v-dobrev @tzanio 

Reminder: Some limited changes were necessary in MFEM package to get this working. In parallel, MFEM package is being updated following the release (@v-dobrev's work in Spack: https://github.com/spack/spack/pull/19748). I am a bit late on the latest changes.

We have 2 options:
- keep going with the branch as-is.
- keep updating the MFEM package in MFEM-Uberenv and wait for merge in Spack to point to the corresponding commit.

**Comment by adrienbernede** [comment]
Ultimately, the local changes in MFEM package (in MFEM-Uberenv) will be merged in Spack to have only one version everywhere. Only the transition is non trivial.

**Comment by adrienbernede** [comment]
@v-dobrev, when do you think the new MFEM Spack package will be merge into Spack?

**Comment by tzanio** [comment]
@acfisher and @cjoycd, is one of you available to provide a second review for this?

**Comment by cjvogl** [comment]
@tzanio and @v-dobrev, what is the due date for this review?  Seems like the original approval date was some time ago...

**Comment by tzanio** [comment]
I'll update the review date.

**Comment by adrienbernede** [comment]
LGTM

File: .gitlab/quartz.yml
Comment (acfisher) [inline .gitlab/quartz.yml:92]: Is this commented code being kept here for a reason?

**Review by acfisher — APPROVED** [review]

File: .gitlab/quartz.yml
Comment (jamiebramwell) [reply .gitlab/quartz.yml:92]: I agree. I assume this is to eventually test our special flavor of builds, but I think keeping the file clean for now is preferred.

**Review by jamiebramwell — APPROVED** [review]
Awesome! I think this will really help catch stuff in our build toolchain.

**Comment by adrienbernede** [comment]
LGTM

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Thanks @adrienbernede !

**Pull request merged**
